File: Python3/2379_minimum_recolors_to_get_k_consecutive_black_blocks.py
# https://leetcode.cn/problems/minimum-recolors-to-get-k-consecutive-black-blocks/

class Solution:
    def minimumRecolors(self, blocks, k):
        """
        Inputs:
            blocks: str
            k: int
        Outputs:
            res: int
        """
        # Sliding window.
        cnt = 0
        for i in range(k):
            if blocks[i] == "B":
                cnt += 1
        res = k - cnt
        for i in range(k, len(blocks)):
            cur, drop = blocks[i], blocks[i-k]
            if cur == "B" and drop == "W":
                cnt += 1
                res = min(res, k - cnt)
            elif cur == "W" and drop == "B":
                cnt -= 1
        return res

File: Cpp/2379_minimum_recolors_to_get_k_consecutive_black_blocks.cpp
// https://leetcode.cn/problems/minimum-recolors-to-get-k-consecutive-black-blocks/

class Solution {
public:
    int minimumRecolors(string blocks, int k) {
        // Sliding window.
        int cnt = 0;
        for (int i = 0; i < k; ++i) {
            if (blocks[i] == 'B') {
                cnt += 1;
            }
        }
        int res = k - cnt;
        for (int i = k; i < blocks.size(); ++i) {
            char cur = blocks[i], drop = blocks[i-k];
            if (cur == 'B' && drop == 'W') {
                cnt += 1;
                res = min(res, k - cnt);
            }
            else if (cur == 'W' && drop == 'B') {
                cnt -= 1;
            }
        }
        return res;
    }
};
